src/value.rs

```diff
@@ -922,3 +981,280 @@ impl Value {
+
+    /// Panic if array index out of bounds. If key is not already in the object,
+    /// insert it with a value of null. Panic if Value is a type that cannot be
+    /// indexed into, except if Value is null then it can be treated as an empty
+    /// object.
+    #[doc(hidden)]
+    fn index_or_insert<'v>(&self, v: &'v mut Value) -> &'v mut Value;
+}
+
+impl Index for usize {
+    fn index_into<'v>(&self, v: &'v Value) -> Option<&'v Value> {
+        match *v {
+            Value::Sequence(ref vec) => vec.get(*self),
+            Value::Mapping(ref vec) => vec.get(&Value::I64(*self as i64)),
+            _ => None,
+        }
+    }
+    fn index_into_mut<'v>(&self, v: &'v mut Value) -> Option<&'v mut Value> {
+        match *v {
+            Value::Sequence(ref mut vec) => vec.get_mut(*self),
+            Value::Mapping(ref mut vec) => vec.get_mut(&Value::I64(*self as i64)),
+            _ => None,
+        }
+    }
+    fn index_or_insert<'v>(&self, v: &'v mut Value) -> &'v mut Value {
+        match *v {
+            Value::Sequence(ref mut vec) => {
+                let len = vec.len();
+                vec.get_mut(*self)
+                    .unwrap_or_else(
+                        || {
+                            panic!(
+                                "cannot access index {} of YAML array of length {}",
+                                self,
+                                len
+                            )
+                        },
+                    )
+            },
+            Value::Mapping(ref mut map) => {
+                // TODO: use entry() once LinkedHashMap supports entry()
+                // https://github.com/contain-rs/linked-hash-map/issues/5
+                if !map.contains_key(&Value::I64(*self as i64)) {
+                    map.insert(Value::I64(*self as i64), Value::Null);
+                }
+                map.get_mut(&Value::I64(*self as i64)).unwrap()
+            },
+            _ => panic!("cannot access index {} of YAML {}", self, Type(v)),
+        }
+    }
+}
+
+impl Index for Value {
+    fn index_into<'v>(&self, v: &'v Value) -> Option<&'v Value> {
+        match *v {
+            Value::Mapping(ref map) => map.get(&self),
+            _ => None,
+        }
+    }
+    fn index_into_mut<'v>(&self, v: &'v mut Value) -> Option<&'v mut Value> {
+        match *v {
+            Value::Mapping(ref mut map) => map.get_mut(&self),
+            _ => None,
+        }
+    }
+    fn index_or_insert<'v>(&self, v: &'v mut Value) -> &'v mut Value {
+        if let Value::Null = *v {
+            let mut map = Mapping::new();
+            map.insert(self.clone(), Value::Null);
+            *v = Value::Mapping(map);
+        }
+        match *v {
+            Value::Mapping(ref mut map) => {
+                // TODO: use entry() once LinkedHashMap supports entry()
+                // https://github.com/contain-rs/linked-hash-map/issues/5
```

Comment:
Thanks for calling this out. I filed #58 to follow up in serde_yaml and https://github.com/serde-rs/json/pull/306 to fix this in serde_json.